example/16_gemm_reduce/gemm_reduce_xdl_common.hpp

```diff
@@ -0,0 +1,424 @@
+auto run_gemm_reduce_max_xdl(ck::index_t M,
+                             ck::index_t N,
+                             ck::index_t K,
+                             ck::index_t StrideA,
+                             ck::index_t StrideB,
+                             ck::index_t StrideC,
+                             bool do_verification,
+                             int init_method,
+                             bool time_kernel)
+{
+    auto f_host_tensor_descriptor =
+        [](std::size_t row, std::size_t col, std::size_t stride, auto layout) {
+            if(std::is_same<decltype(layout), ck::tensor_layout::gemm::RowMajor>::value)
+            {
+                return HostTensorDescriptor(std::vector<std::size_t>({row, col}),
+                                            std::vector<std::size_t>({stride, 1}));
+            }
+            else
+            {
+                return HostTensorDescriptor(std::vector<std::size_t>({row, col}),
+                                            std::vector<std::size_t>({1, stride}));
+            }
+        };
+
+    Tensor<ADataType> a_m_k(f_host_tensor_descriptor(M, K, StrideA, ALayout{}));
+    Tensor<BDataType> b_k_n(f_host_tensor_descriptor(K, N, StrideB, BLayout{}));
+
+    Tensor<CDataType> c_m_n_host_result(f_host_tensor_descriptor(M, N, StrideC, CLayout{}));
+    Tensor<ReduceDataType> reduce_m_host_result(
+        HostTensorDescriptor(std::vector<std::size_t>({static_cast<std::size_t>(M)})));
+
+    Tensor<CDataType> c_m_n_device_result(f_host_tensor_descriptor(M, N, StrideC, CLayout{}));
+    Tensor<ReduceDataType> reduce_m_device_result(
+        HostTensorDescriptor(std::vector<std::size_t>({static_cast<std::size_t>(M)})));
+
+    std::cout << "a_m_k: " << a_m_k.mDesc << std::endl;
+    std::cout << "b_k_n: " << b_k_n.mDesc << std::endl;
+    std::cout << "c_m_n: " << c_m_n_host_result.mDesc << std::endl;
+    std::cout << "reduce_m: " << reduce_m_host_result.mDesc << std::endl;
+
+    switch(init_method)
+    {
+    case 0: break;
+    case 1:
+        ck::utils::FillUniformDistributionIntegerValue<ADataType>{}(a_m_k.begin(), a_m_k.end());
+        ck::utils::FillUniformDistributionIntegerValue<BDataType>{}(b_k_n.begin(), b_k_n.end());
+        break;
+    default:
+        ck::utils::FillUniformDistribution<ADataType>{0.f, 1.f}(a_m_k.begin(), a_m_k.end());
+        ck::utils::FillUniformDistribution<BDataType>{0.5f, 0.5f}(b_k_n.begin(), b_k_n.end());
+        break;
+    }
+
+    DeviceMem a_device_buf(sizeof(ADataType) * a_m_k.mDesc.GetElementSpaceSize());
+    DeviceMem b_device_buf(sizeof(BDataType) * b_k_n.mDesc.GetElementSpaceSize());
+    DeviceMem c_device_buf(sizeof(CDataType) * c_m_n_device_result.mDesc.GetElementSpaceSize());
+    DeviceMem reduce_device_buf(sizeof(ReduceDataType) *
+                                reduce_m_device_result.mDesc.GetElementSpaceSize());
+
+    a_device_buf.ToDevice(a_m_k.mData.data());
+    b_device_buf.ToDevice(b_k_n.mData.data());
+
```

Comment:
So for this way of initializing the data, the higher half of each byte will be `zero` right?  Since `int4_t` only uses the lower half of the byte.   

Another issue,  I guess the host GEMM codes accesses each `int4_t` value as scalar type `int4_t`, which occupies one byte;  but how the GPU kernel codes access the `int4_t` values,  does it assume `int4_t` values are packed in `int8_t` and accessed through `int8_t` vector ?

Reply:
I think we need solve the following here:

1.  Whether the Host GEMM for verification need to support accessing of  packed `int4_t` ?
2.  Initializing the `int4_t` values as packed in8_t values in temporary host buffer before copying them to device memory,  so that GPU kernel  access actually packed int4_t data
3.  When doing checking error,  do have consideration whether the result by the Host GEMM is packed or not.

Reply:
As for if half of byte is zero it looks like it's not: [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEEgCspAAOqAqETgwe3r7%2ByanpAiFhkSwxcVyJdpgOGUIETMQEWT5%2BAVU1AnUNBEUR0bEJtvWNzTltwz2hfaUDFQCUtqhexMjsHOYAzKHI3lgA1CYbbgReSfSH2CYaAIKb27uYB0dejrSEAJ4XV7dmWww7Xn2hzcTgUBGImFYXxud3%2BDyebmIhnQqBY0J%2BfwBQKOThYhjwSXRsKxj2BYmAJEICDRG0uMN%2B90BpKOADdqkRiESGXCmQiCO8kpgAPrgpiEBRczHw4HIPEEBCSxnYtxRcWYAhEm5eNJGPahAiSEVPAAieyFFkIAElBBBJHNDlYYTcwUxHMg9Qw3mEPQajUwogoIPrDQQ9qp7TCAOyO657OM%2BkN7BTABgmsNPS60va2vYAWj2XHtGxj8b2EJOxFTEFUB3i2CTKbmeYbDAd3xMkeN7Zu%2Br2eNCEBZqDw6Ajt2j3dj8bB6BAKCWoeBwIOZjMaQAXphUFQg4I7SAV2YEcuN1ud8GRU2l0ckwRZyBXLQ206p3GZ3O0C9jzfzGYL6GAFkBzmA9f2/Nxb3vLB3WvCCXTdIVRDBYF3xAfUNjMEULl3X0CAgXNIzmMdSxIkjYMPUg9gQTB1EPcDILnaia3I1CFAQVAAHcoiYJR6PgvBkEQniNSOVD9S4AA2LDaRwkN8MIq8jhPO850fZ9bhuUtUM/RclJ/Vd/z2ACmFUCAQLoliVJAaC%2BPqBCkJEtwxMEDDpOwWSRQgBTvlI3z6N/SimNA1dbPvJjQrnNjOO43jLNdAShOQ0SrPEqSRPc/8vKIiKHwYdAn2LSctKsnSctVAglEcgAOL5o34wSHJQqyvH1CSQ2w/1Aw8vDc3iIiiI7U1LPvNTCs06cSoXMq1Wq2qLHqxLHNQlrBDatyoEyhT7U7HLRpLCb71K4a53KyrgRq2kO3muyEsa5L7xWgg1vSja908gj%2BsG3a8oKmMfPjctllTDR1MG74OAWWhOHiXg/A4LRSFQTg3GsawkyWFZSV%2BHhSAITQIYWABrEB4g0fROEkWH8cRzheAUEAybx%2BGIdIOBYCQGjqheEhyEoBpgAUZRDEwWghHYji4ZxtAWCSOh4oEIWwlF8W4YR6XZfoOIdkMYApA0Mn1boWIAHkXjFzjVd4TnkGuYgBdpwJVGqOp8Dh3h%2BEEEQxHYKQZEERQVHUZnSF0Lh9B1lBUcsfQ8ADeAFlQJJHAEemODpjHVj0GdQkVkXzYlzgcfBTA1hxjikSSQuWahjgYdIS2aY4bAneQbniDDKqJNzNq9m13UpAAOg0IfsxRyxrEo3BCBIFcNkLXgma0IjSGopgsDiMzSGJ0nyY4Sn6%2BppG09sBncfx5ea7MKng6Phfz4WNliDSZxJCAA%3D%3D%3D)

Reply:
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/718efa56537db121d1b19f944e6ef0451c30c277/include/ck/tensor_operation/gpu/block/blockwise_gemm_xdlops.hpp#L299-L310 

The GPU GEMM kernel access data as MFMAType which is int8 for int4 input data type. Thus we convert input int4 data in registers to int8 data. Then normally it access data as vector type and runs MFMA instruction on it.